### ISSUE TITLE
block candidate may communicate sigops as sigchecks post fork in coinbase tx

### ISSUE THREAD

**Issue by gandrewstone** [Bitcoin-ABC/bitcoin-abc#379]
### Describe the issue
https://github.com/Bitcoin-ABC/bitcoin-abc/blob/master/src/miner.cpp#L237

After the may2020 fork coinbase tx should report 0 sigchecks, rather than the # of sigops in their output scripts.

I think that this field is output in getblocktemplate as "sigops".  However, since I work on a different client, what you do with the field may have changed, potentially making this not-a-bug.  If that is the case, sorry for wasting your time.

**Comment by Greg-Griffith** [comment]
andrew is right, there is a small bug here. 

in the spec it states 
```
The essential idea of SigChecks is to perform counting solely in the spending 
transaction, and count actual executed signature check operations. 
```
if we are now only counting in the spending transaction, and the coinbase never spents inputs. then it should always have a sigcheck count of 0. 

Your current code here: https://reviews.bitcoinabc.org/source/bitcoin-abc/browse/master/src/miner.cpp$237 
uses the GetSigOpCountWithoutP2SH function which includes outputs
reference: https://reviews.bitcoinabc.org/source/bitcoin-abc/browse/master/src/consensus/tx_verify.cpp$160
so you are reporting the wrong count for the coinbase transaction which should always be 0.

**Comment by markblundeberg** [comment]
You're right -- I realized this a month ago however I am not working on ABC anymore so I haven't done any work on fixing it. The bug is at least mild since it means the block template constructor will *over*estimate the true sigchecks count and thus the only downside will be that it includes slightly fewer normal transactions.

**Comment by gandrewstone** [comment]
Yes, this is very mild.  But, qualified by I'm mostly looking at BU, and just skimmed ABC to see that its similar:  the block is already constructed so it won't affect the true sigchecks count.  This is just a value that gets output to the miner in the getblocktemplate call.  So it won't include slightly fewer transactions.  

The upside is that its just a small thing in a field that miners likely don't use.  The downside is if the miner actually used it, they might sum to > the limit of sigchecks and their code might panic, *except* not in the "normal" p2pkh case because:

I think you may be getting 1 fewer sigcheck because in bool BlockAssembler::TestPackage() the code does:

    if (nBlockSigOps + packageSigOps >=
        MaxBlockSigOpsCountForSize(blockSizeWithPackage)) {
        return false;
    }

As per the spec it should be that > the limit returns false since the spec says <= is valid.  But actually, I chose to keep this same code in the BU code as well just because it is a little more tolerant to not generate up to the exact limit.

**Comment by sickpig** [comment]
> You're right -- I realized this a month ago however I am not working on ABC anymore so I haven't done any work on fixing it. The bug is at least mild since it means the block template constructor will _over_estimate the true sigchecks count and thus the only downside will be that it includes slightly fewer normal transactions.

Mark don't you think that just put a sentence in the spec that explicitly state that coinbase tx's must be counted as 0?

**Comment by ftrader** [comment]
@sickpig:

https://gitlab.com/bitcoin-cash-node/bchn-sw/bitcoincash-upgrade-specifications/-/merge_requests/4
